### PR TITLE
LSPS3: Promise To Unconditionally Fund 0-conf

### PR DESCRIPTION
**Pull request by ZmnSCPxj-jr**

### CODE REVIEW

**Comment by ZmnSCPxj-jr** [comment]
Eh, can just keep LSPS2 empty for some time, not need to renumber everything.

**Comment by ZmnSCPxj-jr** [comment]
As per meeting, TODO: rename to LSPS3 once  Jit channels spec has been put into LSPS2.

**Comment by ZmnSCPxj-jr** [comment]
TODO: replace "fraud proof" with "misbehaviour proof" / "proof of misbehaviour".

**Comment by ZmnSCPxj-jr** [comment]
TODO: should also commit to outnum and amount.

TODO: Describe how client validation can be deferred to closing time rather than funding confirmation time.

**Comment by BitcoinErrorLog** [comment]
Admittedly I didn't have time to read everything, but I wanted to comment on what I did read. I think the idea of using 3rd parties as a punishment may not be as incentive-aligned as you hope, and potentially abusable. 

Having other nodes close all channels and blacklist may not be something that matches the crime, and may come at an undesirable cost to the 3rd party. 

Blacklisting the node notably might really suck if something occurred by mistake.

Further, establishing a credit system is something out of the scope of LN in general, imo. 

The reason we work on Slashtags and Pear Credit, is the greater use cases of identity, context, and credit are not LN-specific problems, and reinventing solutions in LN-specific contexts feels like we run the risk of creating an overly complex monster out the protocol. 

I could be wrong, but my suggestion would be that, if you want to work on separate systems for identity, reputation, or credit problems, I have lots of thoughts and suggestions, but whenever I see this kind of thing baked into LN or Bitcoin, it feels cumbersome and dangerous.

**Comment by ZmnSCPxj-jr** [comment]
>Blacklisting the node notably might really suck if something occurred by mistake.

Something occurring by mistake in this context can mean funds loss for the clients of your LSP.  It seems to me quite commensurate to boycott such a mistake.

**Comment by BitcoinErrorLog** [comment]
I can't see Synonym ever adopting a policy where 3rd parties could affect us into closing a bunch of channels in any automated way. Nor would I want to be ostracized if we accidentally caused such problem. What is your path to redemption in this case? What if the user got refunded and is happy?

I also disagree that the 0conf theft opportunity as a concern for this level of attention warrants this design. LSPs are incentivized NOT to steal. The largest LSP would at best be able to steal from current openings for, what, a day? How many channels do you imagine they are opening and how do you imagine it as worth the risks?

If they are large, the odds are high they are identifiable and accountable, probably even regulated. If they are small, the user has little incentive to use them. 

As I said, I'm not saying that reputation systems, or punishments, are bad ideas outright, I just think baking them directly into LN is resulting in ugly designs and complexity, so far at least.

**Comment by ZmnSCPxj-jr** [comment]
>Nor would I want to be ostracized if we accidentally caused such problem. What is your path to redemption in this case?

I would argue that if you caused a problem like loss of client funds, you fully deserve to be ostracized, whether you want it or not.

Your path to redemption would be to renew your entire LSP (pick a new pubkey) and restart your social credibility from 0 (or negative, most likely).

>What if the user got refunded and is happy?

If they were really happy they would not publish the proof-of-misbehaviour in the first place.

>The largest LSP would at best be able to steal from current openings for, what, a day?

The more pernicious problem is an LSP that randomly steals a small percent of openings for several months, hiding behind "maybe you have Internet connectivity problems or whatever".  Sure there will be isolated reports of "my money is being lost WTF".  The LSP can easily hire excellent PR from the continuous theft to counter this and claim it as FUD by smart-aleck enthusiasts who are compiling their own client and screwing up.

With this specification in place, that kind of slow theft is provable, which forces LSPs to only use exit scams like suddenly stealing all current openings for a day and then vanishing without a trace, which as you noted may not be lucrative enough compared to an ongoing business.  In that case this spec has successfully done its job in stopping a potential theft vector.  See the last section of the specification.

>If they are small, the user has little incentive to use them.

Which is why this spec is valuable.  Everything has to start out small.  If you immediately kill small competitors, you are effectively pushing for a cartelized or even monopolistic economy, which sucks for the users.  There are reasons why laws are often imposed to protect small businesses from larger ones that might want to nip them in the bud.

With this, even if an LSP is small, they can still provide a promise that is as good as a promise from a large LSP.  Yes, they have less to lose by cheating (they likely have fewer and smaller channels open) but they still lose everything they need to operate as a business, and ultimately that is the important thing: to punish in proportion to the size of the criminal.  In human terms, jail time is a punishment in proportion to the wealth of the criminal, as generally the life expectancy of most humans do not vary too much, and thus jail time removes enjoyment of wealth in proportion to the wealth of the criminal.

---

I strongly disagree with the characterization of this as a reputation system.  What is being shared is proof-of-misbehaviour, not arguments like "I got bad service from this LSP, 0 stars LOL" which are trivial to invent, and are the obvious weakness of reputation systems.  The proof-of-misbehaviour requires that the misbehaving LSP have signed using its node ID privkey, and this signature cannot be forged (except at negligible probability).

If some third party managed to get access to the node ID privkey, sure they can trivially invent proofs-of-misbehaviour, but they can also trivially connect to your node masquerading as the other LSP and then send `shutdown` and `error` messages to your channels (`error` even has a convenient "force-close all our channels" option), which also closes your channels with that other LSP, and is triggered by a third party.

If the other LSP keeps its node ID privkey securely safe, then no third party can generate a proof-of-misbehaviour without the other LSP actually misbehaving.  And if the other LSP is the one misbehaving, this is not a third-party triggering your force closes, it is that other LSP misbehaving that is triggering your force closes of the channels with them (and thus a second-party misbehaviour, just as much as they have a bug in their node software which causes them to force-close their channels with you because of the bug).

The protocol is cryptographically secure and the messages cannot be trivially invented without a security breach on the LSP (and if there is a security breach on the LSP, actual funds can be stolen, not just annoy the LSP by closing channels and forcing the LSP to rotate its node ID).

----

More generally, I see two sets of thinking:

* The client should trust the LSP, because the LSP is a big target and can be trusted.
* The client should not trust the LSP, because the LSP is a big guy and can easily beat up the small clients into submission.

In particular, delaying the broadcast of a 0-conf channel funding until after the LSP receives the preimage is a client-should-trust-LSP behaviour, because this is really just the client handing over the preimage without a lien to the actual funds being received (i.e. client hands over preimage "early" before the channel is truly open).  In Lightning this case is normally not trust requiring since the client can always drop onchain and enforce the HTLC, which enforces the atomicity of preimage-transfer+money-transfer.  But if the LSP refuses to broadcast the funding transaction until the client hands over the preimage, then there is no way for the client to actually drop onchain if the LSP then refuses to actually hand over the money.

This specification is from a client-should-not-trust-LSP mindset.

I think this is what really needs to be resolved here.  In some LSPs client-should-not-trust-LSP is considered the ideal: it is OK if the LSP has to trust the client, since most clients are likely to be honest anyway and the LSP is well-positioned to deploy sophisticated thinking to see if a client is less trustworthy then usual.  In other LSPs, client-should-trust-LSP is considered the ideal since the client already has to trust the LSP to allow the client to spend funds and receive them.  Both kinds of LSPs should really need to resolve their differences first.

**Comment by BitcoinErrorLog** [comment]
I do not agree with some of your views on this, including the core belief that clients are ultimately trustworthy. I feel clients trusting LSP is ultimately a safer network than one where LSPs trust clients, else you invite all kinds of attacks, no? 

I think you are right that an obvious proof of cheating could be useful, but how people react to that is up to them. 

Maybe we can first discuss the meta view. Do we want governance like this to be required spec? I have nothing against making various designs and tools, but this spec is speculative in its assumptions, and I'm not sure would be the favored design long term.

**Comment by ZmnSCPxj-jr** [comment]
I do not really see the point of a proof of misbehaviour where "how people react to that is up to them".

Consider that an old commitment transaction getting confirmed is a proof of misbehaviour that is punished by taking the funds of the misbehaving node.  Would it have been better if Lightning were designed so that an old commitment transaction getting confirmed is "how people react to that is up to them" instead of outright punishment?  The punishment mandated here is significantly less punitive, as none of the funds of the misbehaving node is even taken --- it is simply removed from the network, and forced to take up a new identity.

**Comment by ZmnSCPxj-jr** [comment]
Here is the logical problem I am having here:

* On the one hand, people have expressed hesitancy to support "just hand over the preimage and nobody gets hurt".
* On the other hand, other people have expressed hesitancy to support "LSP trusts client".

Again: any mitigation where the LSP defers broadcast of the funding transaction until the client hands over the preimage ***IS*** "just hand over the preimage and nobody gets hurt".  The whole point of designing this PR here is to avoid having to mandate support "just hand over the preimage and nobody gets hurt".

Basically, if we decide to not merge this, I would strongly propose adding the text below to LSPS2:

```Markdown
Clients MUST immediately transmit `channel_ready` after
sending `funding_signed`, and MUST immediately transmit
`update_fulfill_htlc`s as soon as the HTLCs in the payment
have been irrevocably committed on the 0-conf channel.
LSPs SHOULD NOT broadcast the signed funding transaction
until after it has received a `update_fulfill_htlc` for each
HTLC in the payment.
```

Again, the above is basically "just hand over the preimage and nobody gets hurt".

Giving LSPs and clients the option to support one or the other is non-ideal as that leads to subtle incompatibilities when one client requires one thing and the LSP it wants supports only the other.

**Comment by ZmnSCPxj-jr** [comment]
Updated: Change message-to-sign to include "DO NOT MANUALLY SIGN THIS MESSAGE", as per recent LSPS0 common-schemas change.

**Review by SeverinAlexB — COMMENTED** [review]
Summary of my review:
You are trying to build a reputation system for LSPs with misaligned incentives to protect against 0conf double spends. I do not think that this system will work in practice in any way.

Maybe you can disprove my points but I am very pessimistic. Reputation systems are very hard and I see too many issues here.

File: LSPS3/README.md [hunk 1]
```diff
@@ -0,0 +1,691 @@
+# LSPS3: Promise To Unconditionally Fund 0-conf
+
+| Name          | `promise_to_0conf_fund`    |
+| ------------- | -------------------------- |
+| Version       | N / A                      |
+| Status        | For Implementation         |
+
+> **Rationale** Unlike many other LSPS specifications, this protocol does not
+> have any versioning support.
+> TXIDs are deeply ingrained in the consensus rules of the blockchain layer,
```
[the remark this answers is not in the document]
Comment (ZmnSCPxj-jr) [reply LSPS3/README.md:10]: Yes, as all of those use the UTXO model.

File: LSPS3/README.md [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+>   funding transaction for a 0-conf channel.
+>   The client defers releasing the preimage in an `update_fulfill_htlc`
+>   until after it has seen the funding transaction in its local mempool.
+>   The client has the ability to never release the preimage, thus
+>   forcing the LSP to lcok its funds to a channel that is not paid
+>   for, for up to the client-specified `to_self_delay` blocks.
+>   However, the LSP also has the ability to claw back the channel open
+>   by use of an RBF (such as by using a full-RBF fullnode connected to
+>   full-RBF miners) transaction that replaces the funding transaction,
+>   or by otherwise bribing miners.
```
[the remark this answers is not in the document]
Comment (ZmnSCPxj-jr) [reply LSPS3/README.md:85]: Indeed.  This is why this specification exists --- it forces the LSP to get the specific TXID confirmed, or else possibly face punishment.

File: LSPS3/README.md [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+
+The LSP then determines a `target_blockheight` for the funding transaction.
+This is a future blockheight which the LSP absolutely commits to ensuring
+that the funding transaction has been confirmed.
+The LSP is liable for punishment under this specification if it fails to
+have that specific funding transaction ID confirmed before
+`target_blockheight`.
+
+The LSP MUST select a `target_blockheight` that is less than or equal to
+the current blockheight, plus 2016.
```
[the remark this answers is not in the document]
Comment (ZmnSCPxj-jr) [reply LSPS3/README.md:265]: Yes.

**Comment by ZmnSCPxj-jr** [comment]
Withdrawn

**Pull request closed without merging**